nilearn/surface/surface.py

```diff
@@ -1238,19 +1241,308 @@ def check_surface(surface):
+    """A collection of data arrays.
+
+    It is a shallow wrapper around the ``parts`` dictionary, which cannot be
+    empty and whose keys must be a subset of {"left", "right"}.
+
+    Parameters
+    ----------
+    left : :obj:`numpy.ndarray` or :obj:`str` or :obj:`pathlib.Path` or None,\
+           default = None
+
+    right : :obj:`numpy.ndarray` or :obj:`str` or :obj:`pathlib.Path` or None,\
+           default = None
+
+    Attributes
+    ----------
+    parts : :obj:`dict` of :obj:`numpy.ndarray`
+
+    shape : :obj:`tuple` of :obj:`int`
+    """
+
+    def __init__(self, left=None, right=None):
+        if left is None and right is None:
+            raise ValueError(
+                "Cannot create an empty PolyData. "
+                "Either left or right (or both) must be provided."
+            )
+
+        parts = {}
+        if left is not None:
+            if not isinstance(left, np.ndarray):
+                left = load_surf_data(left)
+            parts["left"] = left
+        if right is not None:
+            if not isinstance(right, np.ndarray):
+                right = load_surf_data(right)
+            parts["right"] = right
+
+        if len(parts) == 1:
+            self.parts = parts
+            self.shape = next(iter(self.parts.values())).shape
+            return
+
+        if parts["left"].shape[:-1] != parts["right"].shape[:-1]:
+            raise ValueError(
+                f"Data arrays for keys 'left' and 'right' "
+                "have incompatible shapes: "
+                f"{parts['left'].shape} and {parts['right'].shape}"
+            )
+
+        self.parts = parts
+        first_shape = next(iter(parts.values())).shape
+        concat_dim = sum(p.shape[-1] for p in parts.values())
+        self.shape = (*first_shape[:-1], concat_dim)
+
+    def to_filename(self, filename):
+        """Save data to gifti.
+
+        Parameters
+        ----------
+        filename : :obj:`str` or :obj:`pathlib.Path`
+                   If the filename contains `hemi-L`
+                   then only the left part of the mesh will be saved.
+                   If the filename contains `hemi-R`
+                   then only the right part of the mesh will be saved.
+                   If the filename contains neither of those,
+                   then `_hemi-L` and `_hemi-R`
+                   will be appended to the filename and both will be saved.
+        """
+        filename = _sanitize_filename(filename)
+
+        if "hemi-L" not in filename.stem and "hemi-R" not in filename.stem:
+            for hemi in ["L", "R"]:
+                self.to_filename(
+                    filename.with_stem(f"{filename.stem}_hemi-{hemi}")
+                )
+            return None
+
+        if "hemi-L" in filename.stem:
+            data = self.parts["left"]
+        if "hemi-R" in filename.stem:
+            data = self.parts["right"]
+
+        _data_to_gifti(data, filename)
+
+
+class SurfaceMesh(abc.ABC):
```

Comment:
rename from Mesh to SurfaceMesh to avoid clashing with Mesh named tuple that already exist in this module.